inference-engine/src/mkldnn_plugin/nodes/mkldnn_topk_node.cpp

```diff
@@ -359,54 +1721,53 @@ void MKLDNNTopKNode::topk_axis(const float* src_data, float* dst_data, int* dst_
             max_indexes[index2] = tmp_index;
         };
 
-        for (int i2 = 0; i2 < src_k; i2++) {
+        for (int i2 = 0; i2 < top_k; i2++) {
             max_values[i2] = src_data[s_index];
             max_indexes[i2] = i2;
             s_index += after_num;
         }
-        for (int i2 = 0; i2 < src_k - 1; i2++) {
-            for (int i3 = src_k - 1; i3 > i2; i3--) {
-                if (Compare2<float>()(max_values[i3], max_values[i3 - 1])) {
+        for (int i2 = 0; i2 < top_k - 1; i2++) {
+            for (int i3 = top_k - 1; i3 > i2; i3--) {
+                if (compare(max_values[i3], max_values[i3 - 1])) {
                     swap_func(i3, i3 - 1);
                 }
             }
         }
-        for (int i2 = src_k; i2 < dim; i2++) {
-            max_values[src_k] = src_data[s_index];
-            max_indexes[src_k] = i2;
-            for (int i3 = src_k; i3 > 0; i3--) {
-                if (Compare2<float>()(max_values[i3], max_values[i3 - 1]))
+        for (int i2 = top_k; i2 < dim; i2++) {
+            max_values[top_k] = src_data[s_index];
+            max_indexes[top_k] = i2;
+            for (int i3 = top_k; i3 > 0; i3--) {
+                if (compare(max_values[i3], max_values[i3 - 1]))
                     swap_func(i3, i3 - 1);
                 else
                     break;
             }
             s_index += after_num;
         }
-        if (!sort_value) {
-            for (int i2 = 0; i2 < src_k - 1; i2++) {
-                for (int i3 = src_k - 1; i3 > i2; i3--) {
+        if (sort_index) {
+            for (int i2 = 0; i2 < top_k - 1; i2++) {
+                for (int i3 = top_k - 1; i3 > i2; i3--) {
                     if (std::greater<int>()(max_indexes[i3 - 1], max_indexes[i3])) {
                         swap_func(i3, i3 - 1);
                     }
                 }
             }
         }
         if (dst_data) {
-            for (int i2 = 0; i2 < src_k; i2++)
-                dst_data[i0 * src_k * after_num + i2 * after_num + first_index + i1] = max_values[i2];
+            for (int i2 = 0; i2 < top_k; i2++)
+                dst_data[i0 * top_k * after_num + i2 * after_num + i1] = max_values[i2];
         }
         if (dst_idx) {
-            for (int i2 = 0; i2 < src_k; i2++)
-                dst_idx[i0 * src_k * after_num + i2 * after_num + first_index + i1] = max_indexes[i2];
+            for (int i2 = 0; i2 < top_k; i2++)
+                dst_idx[i0 * top_k * after_num + i2 * after_num + i1] = max_indexes[i2];
         }
     });
 }
 
-template <template <typename> class Compare>
-void MKLDNNTopKNode::topk(const float* src_data, float* dst_data, int* dst_idx, SizeVector in_dims) {
+void MKLDNNTopKNode::topk(const float* src_data, float* dst_data, int32_t* dst_idx, SizeVector in_dims, std::function<float(float, float)> compare) {
```

Comment:
SizeVector can be replaced with const SizeVector&

Reply:
Fixed. Thanks Egor!